crates/arroyo-controller/src/states/mod.rs

```diff
@@ -381,6 +384,102 @@ impl<'a> JobContext<'a> {
             retries: retries.saturating_sub(self.retries_attempted),
         }
     }
+
+    async fn take_stopping_checkpoint(
+        &mut self,
+    ) -> Result<StoppingCheckpointOutcome, StoppingCheckpointError> {
+        let job_controller = self.job_controller.as_mut().unwrap();
+
+        let mut final_checkpoint_started = false;
+        let mut shutdown_started = false;
+
+        loop {
+            match job_controller.checkpoint_finished().await {
+                Ok(done) => {
+                    debug!("checked checkpoint, got {}, job_controller.finished(): {}, final_checkpoint_started: {}", done, job_controller.finished(), final_checkpoint_started);
+
+                    if done && final_checkpoint_started {
+                        if job_controller.finished() {
+                            return Ok(StoppingCheckpointOutcome::SuccessfullyStopped);
+                        } else if !shutdown_started {
+                            info!("Starting shutdown");
```

Comment:
In the controller, logs should use structed logging like https://github.com/ArroyoSystems/arroyo/blob/master/crates/arroyo-controller/src/states/recovering.rs#L24 — this preserves context around job_id and makes it easier to build alerts / monitoring around the log lines